shell/platform/android/io/flutter/embedding/android/FlutterSplashView.java

```diff
@@ -0,0 +1,151 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.android;
+
+import android.content.Context;
+import android.os.Build;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.v4.view.ViewCompat;
+import android.util.AttributeSet;
+import android.view.View;
+import android.widget.FrameLayout;
+
+import io.flutter.embedding.engine.renderer.OnFirstFrameRenderedListener;
+
+/**
+ * {@code View} that displays a {@link SplashScreen} until a given {@link FlutterView}
+ * renders its first frame.
+ */
+public class FlutterSplashView extends FrameLayout {
+  private static String TAG = "FlutterSplashView";
+
+  @Nullable
+  private SplashScreen splashScreen;
+  @Nullable
+  private FlutterView flutterView;
+  @Nullable
+  private View splashScreenView;
+
+  @NonNull
+  private final OnFirstFrameRenderedListener onFirstFrameRenderedListener = new OnFirstFrameRenderedListener() {
+    @Override
+    public void onFirstFrameRendered() {
+      if (splashScreen != null) {
+        splashScreen.transitionToFlutter(onTransitionComplete);
+      }
+    }
+  };
+
+  @NonNull
+  private final Runnable onTransitionComplete = new Runnable() {
+    @Override
+    public void run() {
+      removeView(splashScreenView);
+    }
+  };
+
+  public FlutterSplashView(@NonNull Context context) {
+    super(context);
+  }
+
+  public FlutterSplashView(@NonNull Context context, @Nullable AttributeSet attrs) {
+    super(context, attrs);
+  }
+
+  public FlutterSplashView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
+    super(context, attrs, defStyleAttr);
+  }
+
+  /**
+   * Displays the given {@link SplashScreen} until Flutter renders its first frame.
+   * <p>
+   * The given {@link SplashScreen} is displayed as soon as this method is invoked, or
+   * upon attachment to the {@code Window}, whichever happens last.
+   * <p>
+   * If Flutter has already rendered its first frame, this method does not have any
+   * visual impact.
+   */
+  public void setSplashScreen(@NonNull SplashScreen splashScreen) {
+    removeSplashScreen();
+
+    this.splashScreen = splashScreen;
+
+    if (isAttachedToWindow()) {
+      showSplashScreen();
+    }
+  }
+
+  /**
+   * Displays the given {@link FlutterView} in this {@code FlutterSplashView}, or removes
+   * an existing {@link FlutterView} if {@code flutterView} is {@code null}.
```

Comment:
If the FlutterView hasn't rendered its first frame yet, the splash view will be shown instead.